File: FirebaseDatabase/Sources/Utilities/FStringUtilities.m
/*
 * Copyright 2017 Google
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "FStringUtilities.h"
#import "NSData+SRB64Additions.h"
#import <CommonCrypto/CommonDigest.h>

@implementation FStringUtilities

// http://stackoverflow.com/questions/3468268/objective-c-sha1
// http://stackoverflow.com/questions/7310457/ios-objective-c-sha-1-and-base64-problem
+ (NSString *)base64EncodedSha1:(NSString *)str {
    const char *cstr = [str cStringUsingEncoding:NSUTF8StringEncoding];
    // NSString reports length in characters, but we want it in bytes, which
    // strlen will give us.
    unsigned long dataLen = strlen(cstr);
    NSData *data = [NSData dataWithBytes:cstr length:dataLen];
    uint8_t digest[CC_SHA1_DIGEST_LENGTH];
    CC_SHA1(data.bytes, (unsigned int)data.length, digest);
    NSData *output = [[NSData alloc] initWithBytes:digest
                                            length:CC_SHA1_DIGEST_LENGTH];
    return [FSRUtilities base64EncodedStringFromData:output];
}

+ (NSString *)urlDecoded:(NSString *)url {
    NSString *replaced = [url stringByReplacingOccurrencesOfString:@"+"
                                                        withString:@" "];
    NSString *decoded = [replaced stringByRemovingPercentEncoding];
    // This is kind of a hack, but is generally how the js client works. We
    // could run into trouble if some piece is a correctly escaped %-sequence,
    // and another isn't. But, that's bad input anyways...
    if (decoded) {
        return decoded;
    } else {
        return replaced;
    }
}

+ (NSString *)urlEncoded:(NSString *)url {
    // Didn't seem like there was an Apple NSCharacterSet that had our version
    // of the encoding So I made my own, following RFC 2396
    // https://www.ietf.org/rfc/rfc2396.txt allowedCharacters = alphanum | "-" |
    // "_" | "~"
    NSCharacterSet *allowedCharacters = [NSCharacterSet
        characterSetWithCharactersInString:@"abcdefghijklmnopqrstuvwxyzABCDEFGH"
                                           @"IJKLMNOPQRSTUVWXYZ0123456789-_~"];
    return [url
        stringByAddingPercentEncodingWithAllowedCharacters:allowedCharacters];
}

+ (NSString *)sanitizedForUserAgent:(NSString *)str {
    return
        [str stringByReplacingOccurrencesOfString:@"/|_"
                                       withString:@"|"
                                          options:NSRegularExpressionSearch
                                            range:NSMakeRange(0, [str length])];
}

@end

File: FirebaseDatabaseSwiftCore/Sources/FStringUtilities.swift
//
//  File.swift
//  
//
//  Created by Morten Bek Ditlevsen on 28/10/2021.
//

// NOTE: If deployment target moved to iOS 13 or above, CommonCrypto could be skipped entirely
#if canImport(CommonCrypto)
import CommonCrypto

extension Data {
    func sha1() -> Data {
        // Use CC_SHA1
        var digest = [UInt8](repeating: 0, count:Int(CC_SHA1_DIGEST_LENGTH))
        self.withUnsafeBytes {
            _ = CC_SHA1($0.baseAddress, CC_LONG(self.count), &digest)
        }
        return Data(bytes: digest, count: Int(CC_SHA1_DIGEST_LENGTH))
    }
}
#elseif os(Linux) || os(Windows) || os(Android) || os(FreeBSD)
import Crypto
extension Data {
    func sha1() -> Data {
        // Use Insecure.SHA1
        return Data(Insecure.SHA1.hash(data: self))
    }
}
#endif

import Foundation

@objc public class FStringUtilities: NSObject {
    @objc public static func base64EncodedSha1(_ input: String) -> String {
        FStringUtilitiesSwift.base64EncodedSha1(input)
    }

    @objc public static func urlDecoded(_ url: String) -> String {
        FStringUtilitiesSwift.urlDecoded(url)
    }

    @objc public static func urlEncoded(_ input: String) -> String {
        FStringUtilitiesSwift.urlEncoded(input)
    }
    @objc public static func sanitizedForUserAgent(_ str: String) -> String {
        FStringUtilitiesSwift.sanitizedForUserAgent(str)
    }
}

enum FStringUtilitiesSwift {
    static func base64EncodedSha1(_ input: String) -> String  {
        let data = Data(input.utf8)
        return data.sha1().base64EncodedString()
    }

    static func urlDecoded(_ url: String) -> String {
        let replaced = url.replacingOccurrences(of: "+", with: " ")
        // This is kind of a hack, but is generally how the js client works. We
        // could run into trouble if some piece is a correctly escaped %-sequence,
        // and another isn't. But, that's bad input anyways...
        if let decoded = replaced.removingPercentEncoding {
            return decoded
        } else {
            return replaced
        }
    }

    static let allowedCharacters = CharacterSet(charactersIn: "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-_~")
    static func urlEncoded(_ url: String) -> String {
        // Didn't seem like there was an Apple NSCharacterSet that had our version
        // of the encoding So I made my own, following RFC 2396
        // https://www.ietf.org/rfc/rfc2396.txt allowedCharacters = alphanum | "-" |
        // "_" | "~"
        url.addingPercentEncoding(withAllowedCharacters: allowedCharacters)!
    }

    static func sanitizedForUserAgent(_ str: String) -> String {
        str.replacingOccurrences(of: "/|_",
                                 with: "|",
                                 options: .regularExpression,
                                 range: str.startIndex..<str.endIndex)
    }
}
